### PR TITLE
Feature/undo - observers and metadata

### PR DESCRIPTION
**Pull request by cortfritz**
this PR - currently draft - adds 
- a GenServer for metadata state (as alternative to attempting mutation to metadata across NIF in callbacks)
- methods in undo_manager.ex to add observer callbacks
- methods wrapping yrs in undo.rs to support these callbacks, as well as stack item ID generation (to support identifiability in GenServer state)
- tests to exercise these

to do:
- get and apply feedback from @satoren 
- satisfy CodeRabbit
- test scale and performance
- possibly restructure or rename elixir code
- possibly refactor
- possibly enrich code by exposing more of the underlying yrs implementation
- possibly explore async versions of yrs implementation (currently using blocking)
- possibly add really complex test scenarios

### CODE REVIEW

**Comment by cortfritz** [comment]
another todo: upon merge, submit a change to README for UndoManager support!!!

File: native/yex/src/undo.rs [hunk 1]
```diff
@@ -118,53 +84,77 @@ pub fn undo_manager_new_with_options(
...
+        ..Default::default()
+    };
+
+    let undo_manager =
+        YrsUndoManager::<UndoMetadata>::with_scope_and_options(&doc, &branch, undo_options);
+    let wrapper = UndoManagerWrapper::new(undo_manager);
+    let resource = ResourceArc::new(NifWrap(RwLock::new(wrapper)));
+
+    Ok(resource)
+}
```
[the remark this answers is not in the document]
Comment (cortfritz) [reply native/yex/src/undo.rs]: idk about this.  gets turned immediately into a NifError

File: lib/undo_manager.ex [hunk 2]
```diff
@@ -63,68 +212,228 @@
...
+  @spec on_item_added(t(), item_added_callback()) :: {:ok, t()} | {:error, term()}
+  def on_item_added(%__MODULE__{} = manager, callback) when is_function(callback, 1) do
+    with manager <- ensure_metadata_server(manager),
+         :ok <-
+           Yex.UndoMetadataServer.set_item_added_callback(manager.metadata_server_pid, callback),
+         :ok <-
+           Yex.Nif.undo_manager_observe_item_added(manager.reference, manager.metadata_server_pid) do
+      {:ok, manager}
+    end
+  end
```
[the remark this answers is not in the document]
Comment (cortfritz) [reply lib/undo_manager.ex:249]: i will improve the return value of ensure_metadata_server to be an {:ok, manager} tuple and add a comment to explain that if it doesn't start the server successfully it's not going to return that :ok tuple - no race condition

**Comment by cortfritz** [comment]
ok @satoren i addressed *some* of the coderabbit feedback, felt that other feedback was not the best path.

as noted in the opening post there is plenty of room for ongoing improvement, but I'm submitting this as a usable complete first draft of UndoManager that covers the yjs examples what an UndoManager is supposed to do, and is sufficient for checking the box in your ReadMe for having parity for UndoManager.

i remain wide open for any feedback from you - will address to your satisfaction

**Comment by cortfritz** [comment]
Actually the remaining code rabbit nits are good and I will address tomorrow.

**Comment by cortfritz** [comment]
running stress tests, i've found concurrency errors.  working on this.

@satoren would you like bench and stress tests checked in somewhere?  e.g. under `test/bench/` or under `bench/`
?

File: test/undo_manager_test.exs [hunk 3]
```diff
@@ -855,14 +923,23 @@ defmodule Yex.UndoManagerTest do
...
 
     # Verify initial state
     assert XmlFragment.to_string(xml_fragment) == "Hello<div></div>"
 
     # Undo changes
     UndoManager.undo(undo_manager)
+
+    # Give time for the undo operation to complete
+    # Process.sleep(10)
+
```
[the remark this answers is not in the document]
Comment (cortfritz) [reply test/undo_manager_test.exs:942]: for sleeps: better to delete the sleeps altogether
for child spec: advice won't work, tried various ways to do similar things, won't fix
edge cases: fixed
reduce duplication: once you handle all the SharedTypes it's more complex and harder to read, won't fix
add concurrent operations: won't fix - the solution collides with assert_receive/refute - too complex. instead bench testing with script
others: won't fix, nitty

**Comment by cortfritz** [comment]
@satoren i could use feedback on the second bullet below, but otherwise this is my complete draft of observer.

- i got undo/redo robust using yrs try_undo and try_redo.  you might peek at how i handle this and see if you like it
- `new` and `new_with_options` are still not perfectly reliable.  with a stress of 50 concurrent actors / 1000 total over 30 seconds i probably get 1-2 panics.  this is better than at first, but i'd like to be airtight.  if you have guidance for how to transact against [yrs with_options](https://docs.rs/yrs/latest/src/yrs/undo.rs.html#94) I'd love some direction.  It dies on one of the two unwraps.
- there are more functions in yrs but I think we have enough
- i'd love to figure out how to handle async undo/redo

beyond that, I'll be using this in a solution and will learn more about what pains me and absent other feedback I'll use that to guide future work.

I'm wide open to improvements if you see things you don't love.

**Comment by satoren** [comment]
@cortfritz 
> @satoren would you like bench and stress tests checked in somewhere? e.g. under test/bench/ or under bench/
?

Either is fine, but let’s go with “bench” for the sake of shorter commands.

> * i got undo/redo robust using yrs try_undo and try_redo. you might peek at how i handle this and see if you like it

How about using undo_blocking/redo_blocking, which wait until they can acquire rather than failing during a transaction?
To execute a NIF function asynchronously, you need to spawn a thread within the NIF itself. However, this complicates the execution model, so it is not highly recommended.


> * new and new_with_options are still not perfectly reliable. with a stress of 50 concurrent actors / 1000 total over 30 seconds i probably get 1-2 panics. this is better than at first, but i'd like to be airtight. if you have guidance for how to transact against [yrs with_options](https://docs.rs/yrs/latest/src/yrs/undo.rs.html#94) I'd love some direction. It dies on one of the two unwraps.

Where exactly does the panic occur? Is there a way to reproduce it? I’d like to investigate the cause a bit as well.

**Comment by cortfritz** [comment]
Was using undo blocking and was getting panics. I didn't root cause but have been solid with the undo try strategy. 

Where I'm panicking is in either of the two unwinds in the yrs with_options - the function that does the work creating a new undo manager. 

I'll post the bench scripts that include stress tests and will post instructions for repro. 

Thank you.

**Comment by satoren** [comment]
(This is not limited to the implementation of UndoManager) It seems that blocking inside the nif causes all the schedulers in Beam to stop.
https://github.com/satoren/y_ex/commit/fbf016752bfbde231a7178c3519169faedc01585

There are the following options to resolve this, but it’s a difficult decision:
* Create a thread for each document and perform all read/write operations asynchronously (performance degradation is expected).
* Handle the acquisition or waiting of read/write transactions on the Beam side (performance degradation is expected).
* (A more passive approach) Clearly state in the README as an important piece of information that each Doc can only be used by one process.

**Comment by cortfritz** [comment]
We could offer options: a try interface where yex-library-user can implement logic for retry and a queue/serialization that's simpler to use but has performance tradeoff.

Or we could offer a retry with a timeout param or option, which would be simple.  It did seem to work under some stress.  

if i had to do one, i'd do the latter, because there is no way around it?  we have to serialize somewhere. (?)

can we just stick doc in a genserer?  it's weird because doc is getting all the sideeffects from sharedtype updates.  but the critical functions like `new` could be protected by genserver serialization.

i'm noticing in my stress tests that i was panicking on accessing array out of bounds: i'd pick an item in the array but with 100 actors my index might already be out of bounds by the time i used it.  couldn't figure out how to use doc transactions without locking up.  So I stuck the array in a genserver and the problem went array.

not sure if genservers for everything is right; i'm a lazy coder.

I'm ok with whatever you think and will do my best to update undo manager to your spec.

**Comment by cortfritz** [comment]
crazy idea,...instead of bridging over NIFs, one could port the whole thing to native elixir.  I like the idea of rust level performance.  however, yjs might be about twice as slow as an efficient elixir implementation.  meaning: elixir could be fast enough.  benefit would be the entities being globally addressable.  you could just dump a doc pid into Horde...


crazy, i know.

**Comment by satoren** [comment]
The idea I just came up with is to wrap a single document with a GenServer and have all NIF operations go through it. Operations on things like Text would also be done through the GenServer of that document.

It might get too long, so I’ll continue here.
https://github.com/satoren/y_ex/issues/112

**Comment by cortfritz** [comment]
before i add confusion to the comment thread there, I'm discussing one aspect of this here.  feel free to have me repost on the new issue if you prefer.

let me see if I understand you

consider we have a GenServer, SafeDoc.

in pseudocode:
my_safe_doc = SafeDoc.start_link()

my_text = safe.get_text("my text id string") 

# my_text passes through any calls back to my_safe_doc, which internally passes calls to GenServer, getting the actual yrs text, or actual yrs doc from GenServer state.  

my_text.insert(0, "blah")  # internally calls my_safe_doc passing in "my text id string" to identify which text to get from state, then does the actual insert on that text.

right?

I think in that case the UndoManager work is ported to work with the new GenServer, and is itself a GenServer.

At some point, an option is to remove the rust dependency by porting the entire CRDT to elixir.  Would be slower than the rust, but native elixir and globally addressable (see "crazy idea" above).  and would still be faster than yjs if we implemented correctly.

**Comment by satoren** [comment]
@cortfritz 
> my_text.insert(0, "blah") # internally calls my_safe_doc passing in "my text id string" to identify which text to get from state, then does the actual insert on that text.
right?

Yes.
I am currently working on it in the [this PR](https://github.com/satoren/y_ex/pull/114), and it now supports operations on Array. This should allow the user experience to remain much the same.


Reimplementing everything in Elixir sounds interesting, but it would probably be slower than Rust, and I don’t have the time for it.

**Comment by satoren** [comment]
@cortfritz The issue of concurrent execution was resolved by performing operations on the document in the process where it was created.
